### PR TITLE
vmpu_buffer_is_ok for v7-M

### PR DESCRIPTION
**Pull request by niklas-arm**
- add `vmpu_buffer_is_ok` for v7-M
- fix `vmpu_buffer_is_ok` for v8-M (also check page heap)
- fix reading xpsr from invalid memory on v7-M

cc @Patater

### CODE REVIEW

File: core/vmpu/src/mpu_armv7m/vmpu_armv7m_mpu.c [hunk 1]
```diff
@@ -406,6 +407,16 @@ bool vmpu_buffer_access_is_ok(int box_id, const void * addr, size_t size)
         return false;
     }
 
+    /* Check if addr range lies in page heap. */
```
Comment (Patater) [inline core/vmpu/src/mpu_armv7m/vmpu_armv7m_mpu.c]: If this is going to look the same in each of the MPU drivers, can we make this a function in `page_allocator.c`, callable by the MPU drivers instead?

**Comment by niklas-arm** [comment]
Unit tests pass on K64F.

File: core/debug/src/core_armv7m/debug_box_armv7m.c [hunk 2]
```diff
@@ -41,7 +42,11 @@ void debug_deprivilege_and_return(void * debug_handler, void * return_handler,
     uint8_t dst_id = g_debug_box.box_id;
 
     /* Copy the xPSR from the source exception stack frame. */
-    uint32_t xpsr = vmpu_unpriv_uint32_read((uint32_t) &((uint32_t *) src_sp)[7]);
+    uint32_t * xpsr_p = &((uint32_t *) src_sp)[7];
+    uint32_t xpsr = 0;
+    if (vmpu_buffer_access_is_ok(g_active_box, xpsr_p, 4)) {
```
Comment (Patater) [inline core/debug/src/core_armv7m/debug_box_armv7m.c]: Consider using `sizeof` instead of `4`.

**Comment by Patater** [comment]
`box_id_namespace_unowned_buffer` test is breaking due to the introduction of these changes.

File: core/vmpu/src/mpu_armv8m/vmpu_armv8m_mpu.c [hunk 3]
```diff
@@ -283,6 +284,14 @@ bool vmpu_buffer_access_is_ok(int box_id, const void * addr, size_t size)
         return false;
     }
 
+    /* Check if addr range lies in page heap. */
```
Comment (Patater) [inline core/vmpu/src/mpu_armv8m/vmpu_armv8m_mpu.c]: This should be in its own commit, as it fixes the v8-M implementation.

"armv8m: Check page heap in vmpu_buffer_access_is_ok"

File: core/system/src/page_allocator_faults.c [hunk 4]
```diff
@@ -48,6 +48,23 @@ uint32_t page_allocator_get_faults(uint8_t page)
     return 0;
 }
 
+
+int page_allocator_check_range_for_box(int box_id, uint32_t start_addr, uint32_t end_addr)
```
Comment (Patater) [inline core/system/src/page_allocator_faults.c]: Add this in a separate commit, but use it from the commit "Implement vmpu_buffer_access_is_ok"

It makes sense to have it a part of the following commit, as we add and use the feature in existing code. A separate commit adds the v7-M implementation of buffer_is_ok, considering the page allocator check from the beginning.
"armv8m: Check page heap in vmpu_buffer_access_is_ok"

File: core/vmpu/src/mpu_armv7m/vmpu_armv7m_mpu.c [hunk 5]
```diff
@@ -274,6 +311,11 @@ bool vmpu_region_get_for_box(uint8_t box_id, const MpuRegion * * const region, u
     return false;
 }
 
+static bool value_in_range(size_t start, size_t end, size_t value)
```
Comment (Patater) [inline core/vmpu/src/mpu_armv7m/vmpu_armv7m_mpu.c]: Consider moving `value_in_range` to a header file shared by the vMPU implementations, so they don't each need a copy.

File: core/debug/src/core_armv7m/debug_box_armv7m.c [hunk 6]
```diff
@@ -41,7 +42,11 @@ void debug_deprivilege_and_return(void * debug_handler, void * return_handler,
     uint8_t dst_id = g_debug_box.box_id;
 
     /* Copy the xPSR from the source exception stack frame. */
-    uint32_t xpsr = vmpu_unpriv_uint32_read((uint32_t) &((uint32_t *) src_sp)[7]);
+    uint32_t * xpsr_p = &((uint32_t *) src_sp)[7];
+    uint32_t xpsr = (1UL << 24UL);
```
Comment (Patater) [inline core/debug/src/core_armv7m/debug_box_armv7m.c]: Use `xPSR_T_Msk`.

File: core/debug/src/core_armv7m/debug_box_armv7m.c [hunk 7]
```diff
@@ -41,7 +42,11 @@ void debug_deprivilege_and_return(void * debug_handler, void * return_handler,
     uint8_t dst_id = g_debug_box.box_id;
 
     /* Copy the xPSR from the source exception stack frame. */
-    uint32_t xpsr = vmpu_unpriv_uint32_read((uint32_t) &((uint32_t *) src_sp)[7]);
+    uint32_t * xpsr_p = &((uint32_t *) src_sp)[7];
+    uint32_t xpsr = xPSR_T_Msk;
+    if (vmpu_buffer_access_is_ok(g_active_box, xpsr_p, sizeof(xpsr_p))) {
```
Comment (Patater) [inline core/debug/src/core_armv7m/debug_box_armv7m.c]: Is it more semantically correct to use `sizeof(*xpsr_p)`?
